### ISSUE TITLE
map_partitions or map_blocks with large objects eats up scheduler memory

### ISSUE THREAD

**Issue by rikturr** [dask/dask#7860]
Opening this report from investigation in https://github.com/dask/dask-ml/issues/842 and https://github.com/dask/dask-ml/pull/843

When passing some large objects as arguments to map_blocks / map_partitions, the scheduler memory can quickly be overwhelmed. Large objects _should_ be wrapped in delayed to add them to graph and avoid this issue ([see here](https://github.com/dask/dask/blob/626cc724ffa1d14cd74f48ff4f21ef443d832afa/dask/array/core.py#L428)) , but the way the object sizes are determined miss some large objects. In the cases I've encountered, it does not correctly compute the size of scikit-learn estimators.

This is because sys.getsizeof does not properly traverse object references to compute the "real" size of an object. This is a notoriously difficult thing to do in Python, so not sure what the best course of action here should be. 

**Minimal Complete Verifiable Example**:

Running on machine with 2 cores and 16 GB of RAM

```python
from dask_ml.datasets import make_classification
from sklearn.ensemble import RandomForestClassifier
import numpy as np
import pickle
import sys
import dask
from dask.distributed import Client

client = Client()

X, y = make_classification(
    n_samples=50000,
    chunks=1000,
    random_state=42,
)

rf = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
_ = rf.fit(X, y)


def dask_predict(part, model):
    return model.predict(part)

preds = X.map_blocks(
    dask_predict,
    model=rf,
    dtype="int",
    drop_axis=1,
)
preds.compute()
```

Scheduler memory ballons up to this:

<img width="458" alt="image" src="https://user-images.githubusercontent.com/8964039/124323726-c7727b00-db4f-11eb-9b2b-4ae9f02b282b.png">

And ends up with this error:

```
KilledWorker: ("('normal-dask_predict-4b858b85224825aeb2d45678c4c91d39', 27)", <WorkerState 'tcp://127.0.0.1:33369', name: 0, memory: 0, processing: 50>)
```

If we explictly delayed the rf object, 

```python
rf_delayed = dask.delayed(rf)
preds = X.map_blocks(
    dask_predict,
    model=rf_delayed,
    dtype="int",
    drop_axis=1,
    meta=np.array([1]),
)
preds.compute()
```

the memory use looks like this:

<img width="750" alt="Screen Shot 2021-07-02 at 4 00 52 PM" src="https://user-images.githubusercontent.com/8964039/124323750-d1947980-db4f-11eb-826f-9bf82e334202.png">

**Environment**:

- Dask version: 2021.5.1
- Python version: 3.7
- Operating System: ubuntu
- Install method (conda, pip, source): conda

**Comment by jrbourbeau** [comment]
Thanks for the clear writeup @rikturr! Indeed it looks like we're underestimating the size of the `RandomForestClassifier` model in your example:

```python
In [1]: from sklearn.ensemble import RandomForestClassifier

In [2]: rf = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)

In [3]: from dask.sizeof import sizeof

In [4]: sizeof(rf)
Out[4]: 48
```

One way to improve the situation is to register a custom `sizeof` implementation in [`dask/sizeof.py`](https://github.com/dask/dask/blob/965e35fe5a9423f550991bde08116a4984498e68/dask/sizeof.py) for scikit-learn estimators which more accurately captures the memory footprint of an estimator. 

One thing that comes to mind is to include information from `estimator.get_params()`

```python
In [5]: rf.get_params()
Out[5]:
{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': -1,
 'oob_score': False,
 'random_state': 42,
 'verbose': 0,
 'warm_start': False}

In [6]: sizeof(rf.get_params())
Out[6]: 2604
```

though there may be other attributes which are stored on the class, but not captured by `get_params`. 

cc @thomasjpfan as you might find this interesting

**Comment by rikturr** [comment]
That would be great! I think there is more than in `get_params`, if you pickle the object and check that size it is much larger:

```python
In [2]: sys.getsizeof(rf), sys.getsizeof(pickle.dumps(rf))
Out[2]: (64, 115509472)
```

Another place this comes up is with xgboost and lightgbm, so would want to consider models from those libraries also

**Comment by stsievert** [comment]
I think https://github.com/scikit-learn/scikit-learn/issues/8642 tracks this issue for Scikit-Learn.

**Comment by thomasjpfan** [comment]
`get_params` only shows the parameters in `__init__` which does not contain the fitted attributes in an estimator.

I see a possible way forward to resolve this issue:

1. Ensure all C/C++ objects in scikit-learn defines a `__sizeof__`. For this specific issue, define it for scikit-learn's tree data structure. (This is partly done in https://github.com/scikit-learn/scikit-learn/pull/15526)
2. **Option 2a**: Add a special case for `BaseEstimator` in https://github.com/dask/dask/blob/main/dask/sizeof.py with a `register_lazy` that calls `sizeof` on `estimator.__dict__`.
3. **Extra**: The booster objects in xgboost and lightgbm would also need to define a `__sizeof__`. Since both libraries have a soft dependency on scikit-learn, they will get the benefit from **Option 2a** only when scikit-learn is installed.

**Option 2b**: To support xgboost/lightgbm better, a more generic solution would be to ducktype the estimator:

```python
@sizeof.register(object)
def sizeof_default(o):
    if hasattr(o, "get_params") and callable(o.get_params):
        # "estimator-like"
        return sizeof(o.__dict__)
    return getsizeof(o)
```

(Would have been a little easier with [Protocol](https://docs.python.org/3/library/typing.html#typing.runtime_checkable) but that requires Python>=3.8)

What does everyone think of either **Option 2a** or **2b**?

**Comment by jsignell** [comment]
I prefer **Option 2b**. It seems like a good idea to do as well as we can with the `sizeof` estimation.
